hathor/p2p/sync_checkpoints.py

```diff
@@ -0,0 +1,213 @@
+        """Iterate over checkpoints_to_sync and find a valid interval to sync, pruning already synced intervals.
+
+        Will only return None when there are no more intervals to sync.
+        """
+        # for checkpoint in self.checkpoints_to_sync[::]:
+        for checkpoint in self.checkpoints_to_sync[::-1]:
+            sync_interval = self._checkpoint_sync_interval(checkpoint)
+            if sync_interval is not None:
+                return sync_interval
+            else:
+                self.checkpoints_to_sync.remove(checkpoint)
+        return None
+
+    def stop(self):
+        self.log.debug('stop sync')
+        self.is_running = False
+
+    def try_to_sync(self):
+        self.log.debug('try to sync')
+
+        if self.peer_syncing:
+            self.log.debug('already syncing')
+            return
+
+        if not self.is_running:
+            self.log.debug('not running')
+            return
+
+        if self.peers_to_request:
+            self.peer_syncing = self.peers_to_request[0]
+            self.run_sync()
+        else:
+            self.log.debug('no peers to sync from, try again later')
+            self.manager.reactor.callLater(5, self.try_to_sync)
+
+    def continue_sync(self) -> None:
+        self.peer_syncing = None
+        self.manager.reactor.callLater(0, self.try_to_sync)
+
+    def run_sync(self):
+        if not self.is_running:
+            self.log.debug('not running')
+            return
+
+        if self.manager.tx_storage.has_needed_tx() and not self.should_skip_sync_tx:
+            # Streaming ended. If there are needed txs, prioritize that
+            return self.peer_syncing.run_sync_transactions()
+
+        if not self.checkpoints_to_sync:
+            self.log.debug('no checkpoints to sync')
+            return
+
+        sync_interval = self._get_next_sync_interval()
+        if not sync_interval:
+            self.log.debug('no checkpoints to sync anymore')
+            return
+
+        self.peer_syncing.run_sync_between_heights(*sync_interval)
+
+        # XXX: reset skip flag
+        self.should_skip_sync_tx = False
+
+    def sync_error(self):
+        self.log.debug('sync error')
+        # Send peer_syncing to banned_peers and start again with another peer
+        self.banned_peers.append(self.peer_syncing)
+        self.peers_to_request.remove(self.peer_syncing)
+        self.peer_syncing = None
+        self.manager.reactor.callLater(0, self.try_to_sync)
+
+    def sync_ended(self):
+        self.log.debug('sync ended')
+
+        if self.manager.tx_storage.has_needed_tx():
+            self.continue_sync()
+            return
+
+        if not self.checkpoints_to_sync:
+            return
+
+        # Double check I have the first checkpoint
+        first_cp = self.checkpoints_to_sync[0]
+
+        if not self.manager.tx_storage.transaction_exists(first_cp.hash):
+            # the sync ended but I still don't have the checkpoint
```

Comment:
Why would it happen?

Reply:
There is a typo, I'll fix it, it's supposed to say "the stream ended", this is what happens when a stream of blocks ends, several streams are usually needed to reach a checkpoint.

Reply:
Moved to another PR.